arangod/Cluster/ClusterInfo.cpp

```diff
@@ -2465,76 +2491,145 @@ int ClusterInfo::ensureIndexCoordinator(
   if (application_features::ApplicationServer::isStopping()) {
     return errorCode;
   }
+  loadPlan();
+
+  // Smart collection, we're done.
+  auto const resultSlice = resultBuilder.slice();
+  if (resultSlice.hasKey("isSmart") &&
+      resultSlice.get("isSmart").isBoolean() &&
+      resultSlice.get("isSmart").getBoolean()) {
+    return TRI_ERROR_NO_ERROR;
+  }
   
-  if (errorCode == TRI_ERROR_HTTP_PRECONDITION_FAILED) { // nothing to rollback
-    AgencyComm ac;
-    AgencyCommResult ag = ac.getValues("/");
-    if (ag.successful()) {
-      LOG_TOPIC(ERR, Logger::CLUSTER) << "Agency dump:\n" << ag.slice().toJson();
-    } else {
-      LOG_TOPIC(ERR, Logger::CLUSTER) << "Could not get agency dump!";
-    }
-    return errorCode;
+  std::string const indexPath =
+    "Plan/Collections/" + databaseName + "/" + collectionID + "/indexes";
+
+  if (slice.get(
+        arangodb::StaticStrings::IndexType).isEqualString("arangosearch")) {
+    return TRI_ERROR_NO_ERROR;
   }
 
-  std::shared_ptr<VPackBuilder> planValue;
-  std::shared_ptr<VPackBuilder> oldPlanIndexes;
-  std::shared_ptr<LogicalCollection> c;
+  VPackBuilder oldPlanIndex;
+  { VPackObjectBuilder b(&oldPlanIndex);
+    for (auto const& entry : VPackObjectIterator(resultSlice)) {
+      auto const key = entry.key.copyString();
+      if (key != "isNewlyCreated") {
+        oldPlanIndex.add(entry.key.copyString(), entry.value);
+      }}}
+  
+  if (errorCode == TRI_ERROR_NO_ERROR) { // Success so far
 
-  size_t tries = 0;
-  do {
-    loadPlan();
-    // find index in plan
-    planValue = nullptr;
-    oldPlanIndexes.reset(new VPackBuilder());
-
-    c = getCollection(databaseName, collectionID);
-    c->getIndexesVPack(*(oldPlanIndexes.get()), Index::makeFlags(Index::Serialize::Basics));
-    VPackSlice const planIndexes = oldPlanIndexes->slice();
-
-    if (planIndexes.isArray()) {
-      for (auto const& index : VPackArrayIterator(planIndexes)) {
-        auto idPlanSlice = index.get(StaticStrings::IndexId);
-        if (idPlanSlice.isString() && idPlanSlice.isEqualString(idString)) {
-          planValue.reset(new VPackBuilder());
-          planValue->add(index);
+    std::string idStr;
+    if (resultSlice.hasKey("id")) {
+      idStr = resultSlice.get("id").copyString();
+    } else {
+      LOG_TOPIC(ERR, Logger::CLUSTER)
+        << "Missing 'id' field in index creation result";
+      return TRI_ERROR_INTERNAL;
+    }
+
+    // isBuilding key to be removed, when index appears in all shards in Current
+    if (resultSlice.hasKey("isBuilding")) {
+      
+      VPackBuilder newPlanIndex;
+      std::uint64_t sleepFor = 50;
+      Result current;
+      
+      while(true) { // Wait for index to appear in current shards
+        current =
+          getCollectionCurrent(databaseName, collectionID)->hasIndex(idStr);
```

Comment:
Is it clear that `getCollectionCurrent` cannot return a `nullptr` here?

Reply:
cannot return nullptr. but a collection with 0 version. handled now.